index.js

```diff
@@ -137,11 +138,58 @@ class Installer {
           this.pkgCount++
           return this
         })
+        .catch(e => {
+          if (child.optional) {
+            this.pkgCount++
+            child.failed = true
+            return rimraf(childPath)
+          } else {
+            throw e
+          }
+        })
       })
       return child.pending
     }, { concurrency: 50 })
   }
 
+  // A cute little mark-and-sweep collector!
+  garbageCollect (tree) {
+    const liveDeps = new Set()
+    const installer = this
+    mark(tree)
+    return sweep(tree)
+
+    function mark (tree) {
+      for (let dep of tree.dependencies.values()) {
+        if (dep.seenByMark) { continue }
+        dep.seenByMark = true
+        if (!dep.optional || !dep.failed) {
```

Comment:
At this point, shouldn’t the only failed deps be optional, otherwise they would have thrown in the extractTree phase before this?

Reply:
I can probably get rid of that `dep.optional`. I think it's an artifact of an experiment, and pretty much useless now.

Reply:
This seems like the only piece of feedback to incorporate.